pkg/southbound/synchronizer/deviceUpdate.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2020-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package synchronizer
+
+import (
+	"strconv"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/cenkalti/backoff"
+
+	"github.com/onosproject/onos-config/pkg/events"
+
+	topodevice "github.com/onosproject/onos-topo/api/device"
+)
+
+func (s *Session) updateDevice(id topodevice.ID, connectivity topodevice.ConnectivityState, channel topodevice.ChannelState,
+	service topodevice.ServiceState) error {
+	log.Info("Update device state")
+
+	topoDevice, err := s.deviceStore.Get(id)
+
+	st, ok := status.FromError(err)
+
+	// If the device doesn't exist then we should not update its state
+	if ok && err != nil && st.Code() == codes.NotFound {
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	protocolState, index := containsGnmi(topoDevice.Protocols)
+	if protocolState != nil {
+		topoDevice.Protocols = remove(topoDevice.Protocols, index)
+	} else {
+		protocolState = new(topodevice.ProtocolState)
+	}
+
+	protocolState.Protocol = topodevice.Protocol_GNMI
+	protocolState.ConnectivityState = connectivity
+	protocolState.ChannelState = channel
+	protocolState.ServiceState = service
+	topoDevice.Protocols = append(topoDevice.Protocols, protocolState)
+	mastershipState, err := s.mastershipStore.GetMastership(topoDevice.ID)
+	if err != nil {
+		return err
+	}
+
+	if topoDevice.Attributes == nil {
+		topoDevice.Attributes = make(map[string]string)
+	}
+
+	topoDevice.Attributes[mastershipTermKey] = strconv.FormatUint(uint64(mastershipState.Term), 10)
+	_, err = s.deviceStore.Update(topoDevice)
```

Comment:
IIRC the device service will reject this update if the device revision has changed. This is the concurrency control that ensures two nodes cannot overwrite each other’s changes. But since those updates could occur in any order (an older master might win the race, in which case the newer master’s write will be rejected), this entire read-modify-write sequence has to be retried until one of two conditions is met:
* The update is successful or
* Upon retry, the node encounters a mastership term greater than its own